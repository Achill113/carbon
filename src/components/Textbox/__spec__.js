import React from 'react';
import TestUtils from 'react/lib/ReactTestUtils';
import Textbox from './index';

describe('Textbox', () => {
  it('failing test demo', () => {
    expect(1).toEqual(2);
  });
});
