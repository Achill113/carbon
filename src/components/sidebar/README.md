# A Sidebar widget.

## How to use a Sidebar in a component:

* In your file

```javascript
  import Sidebar from 'carbon/lib/components/sidebar';
```

* To render the Sidebar:

```javascript
  <Sidebar
    onClose={ closeSidebar }
    open={ true }
  />
```

Side bar is positioned on the right hand screen of the window by default.
To position the sidebar on the left hand side pass `position='left' to the component.

The background behind the sidebar is disabled by default. To allow the user to interact
with all the UI pass `disableBackground={ false }` to the component

| Name              | Required    | Type           | Default       | Description   |
| -------------     | ----------- | -------------  | ------------- | ------------- |
| onClose           | true        | Function       |               | Callback for when close icon is clicked |
| open              | true        | Boolean        | `false`       | Sets the open state of the sidebar      |
| disableBackground | false       | Boolean        | `true`        | Disables background behind sidebar      |
| onClose           | false       | Boolean        | `right`       | 'left' or 'right' position of sidebar   |
