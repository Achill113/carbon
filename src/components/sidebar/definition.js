import { Sidebar } from './';
import OptionsHelper from './../../utils/helpers/options-helper';
import Definition from './../../../demo2/utils/definition';
import sidebarHeaderDefinition from './sidebar-header/definition';

let definition = new Definition('sidebar', Sidebar, {
  description: '[content needed] Basic example of the component',
  designerNotes: '[content needed] Basic designs description for the component',
  associatedDefinitions: [sidebarHeaderDefinition],
  propOptions: {
    position: OptionsHelper.alignBinary,
    size: OptionsHelper.sizesFull
  },
  propValues: {
    children: `<SidebarHeader>
    Header Content
  </SidebarHeader>

  Main Content`
  },
  propTypes: {
    position: "String",
    size: "String"
  },
  propDescriptions: {
    position: "Sets the position of sidebar, either left or right.",
    size: "Sets the size of the sidebar when open."
  }
});

definition.isAModal();

export default definition;
