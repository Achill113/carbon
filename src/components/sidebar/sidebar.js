import React from 'react';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import Icon from './../icon';
import classNames from 'classnames';

class Sidebar extends React.Component {

  static propTypes = {

    /**
     * A custom close event handler
     *
     * @property onCancel
     * @type {Function}
     */
    onCancel: React.PropTypes.func.isRequired,

    /**
     * Sets the open state of the dialog
     *
     * @property open
     * @type {Boolean}
     * @default false
     */
    open: React.PropTypes.bool.isRequired,

    /**
     * Determines if the background is disabled
     * when the sidebar is open
     *
     * @property disableBackground
     * @type {Boolean}
     * @default true
     */
    disableBackground: React.PropTypes.bool,

    /**
     * Determines the position of the sidebar
     * 'left' or 'right'
     *
     * @property position
     * @type {String}
     * @default 'right'
     */
    position: React.PropTypes.string
  }

  static defaultProps = {
    open: false,
    disableBackground: true,
    position: 'right'
  }

  /**
   * Returns classes for the dialog.
   *
   * @method sidebarClasses
   * @return {String} sidebar className
   */
  get sidebarClasses() {
    return classNames(
      'ui-sidebar__sidebar',
      `ui-sidebar__sidebar--${this.props.position}`
    );
  }

  /**
   * Returns classes for the component.
   *
   * @method mainClasses
   * @return {String} Main className
   */
  get mainClasses() {
    return classNames(
      'ui-sidebar',
      this.props.className
    );
  }


  /**
   * Returns HTML for the background.
   *
   * @method backgroundHTML
   * @return {Object} JSX
   */
  get backgroundHTML() {
    if (this.props.disableBackground) {
      return <div className="ui-sidebar__background"></div>;
    }
  }

  /**
   * Returns the computed HTML for the sidebar.
   *
   * @method sidebarHTML
   * @return {Object} JSX for sidebar
   */
  get sidebarHTML() {
    return (
      <div className={ this.sidebarClasses } >
        <Icon className="ui-sidebar__close" type="close" onClick={ this.props.onCancel } />
        { this.props.children }
      </div>
    );
  }

  /**
   * Renders the component.
   *
   * @method render
   * @return {Object} JSX
   */
  render() {
    let backgroundHTML,
        sidebarHTML;

    if (this.props.open) {
      backgroundHTML = this.backgroundHTML;
      sidebarHTML= this.sidebarHTML;
    }

    return (
      <div className={ this.mainClasses }>
        <ReactCSSTransitionGroup
          transitionName={ `sidebar--${this.props.position}` }
          transitionEnterTimeout={ 500 }
          transitionLeaveTimeout={ 500 } >
          { sidebarHTML }
        </ReactCSSTransitionGroup>

        <ReactCSSTransitionGroup
          transitionName="sidebar-background"
          transitionEnterTimeout={ 500 }
          transitionLeaveTimeout={ 500 } >
          { backgroundHTML }
        </ReactCSSTransitionGroup>
      </div>
    );
  }
}

export default Sidebar;
