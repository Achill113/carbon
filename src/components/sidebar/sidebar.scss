@import 'colors';
@import 'mixins';
@import 'z-index';

$sidebar-animation-length: 300ms;
$sidebar-background-opacity: 0.6;
$sidebar-background-animation-length: 300ms;

.ui-sidebar__sidebar {
  background-color: white;
  border-radius: 1px;
  @include box-shadow(-1px 0 0 0 #D8D9DC) 
  position: fixed;
  top: 0;
  right: 0;
  height: 100%;
  width: 400px;
  z-index: $z-sidebar;
}

.ui-sidebar__background {
  background-color: #3C424F;
  bottom: 0;
  left: 0;
  opacity: $sidebar-background-opacity;
  position: fixed;
  right: 0;
  top: 0;
  z-index: $z-sidebar-background;
}

.ui-sidebar__close {
  position: absolute;
  right: 20px;
  top: 15px;
  z-index: 1;

  &:hover {
    color: $blue;
  }
}

.sidebar-enter {
  opacity: 0;
  right: -30px;
}

.sidebar-enter.sidebar-enter-active {
  opacity: 1;
  right: 0;
  transition: all $sidebar-animation-length ease-in;
}

.sidebar-leave {
  right: 0;
  opacity: 1;
}

.sidebar-leave.sidebar-leave-active {
  opacity: 0;
  right: -50px;
  transition: all $sidebar-animation-length ease-in;
}

// sidebar background animation:

.sidebar-background-enter {
  opacity: 0;
}

.sidebar-background-enter.sidebar-background-enter-active {
  opacity: $sidebar-background-opacity;
  transition: opacity $sidebar-background-animation-length ease-in;
}

.sidebar-background-leave {
  opacity: $sidebar-background-opacity;
}

.sidebar-background-leave.sidebar-background-leave-active {
  opacity: 0;
  transition: opacity $sidebar-background-animation-length ease-in;
}
