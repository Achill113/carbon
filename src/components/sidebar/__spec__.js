import React from 'react';
import { shallow, mount, ReactWrapper } from 'enzyme';
import Browser from '../../utils/helpers/browser';
import { Sidebar } from './sidebar';
import Textbox from './../textbox';
import Portal from './../portal';
import Icon from './../icon';

describe('Sidebar', () => {
  let wrapper, leftWrapper, leftPortalContent, portalContent, spy;

  beforeEach(() => {
    spy = jasmine.createSpy();

    wrapper = shallow(
      <Sidebar
        open
        onCancel={ spy }
        className='custom-class'
      >
        <Textbox />
        <Textbox />
        <Textbox />
      </Sidebar>
    );
    portalContent = new ReactWrapper(
      wrapper.find(Portal).prop('children')
    );

    leftWrapper = shallow(
      <Sidebar
        open
        size='small'
        position='left'
        enableBackgroundUI
      />
    );
    leftPortalContent = new ReactWrapper(
      leftWrapper.find(Portal).prop('children')
    );
  });

  describe('componentDidUpdate', () => {
    let mockWindow;
    let wrapper;
    let instance;

    beforeEach(() => {
      mockWindow = {
        addEventListener() {},
        removeEventListener() {}
      };

      spyOn(Browser, 'getWindow').and.returnValue(mockWindow);
    });

    describe('when the Sidebar is open', () => {
      beforeEach(() => {
        wrapper = mount(
          <Sidebar open onCancel={ spy } />
        );
        instance = wrapper.instance();
      });

      it('sets up event listeners to resize and close the Sidebar', () => {
        spyOn(mockWindow, 'addEventListener');
        wrapper.setProps({ title: 'Sidebar title' });

        expect(mockWindow.addEventListener.calls.count()).toEqual(1);
        expect(mockWindow.addEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
      });

      describe('when the Sidebar is already listening', () => {
        it('does not set up event listeners', () => {
          spyOn(mockWindow, 'addEventListener');
          instance.listening = true;
          wrapper.setProps({ title: 'Already listening' });

          expect(mockWindow.addEventListener.calls.count()).toEqual(0);
          expect(mockWindow.addEventListener).not.toHaveBeenCalled();
        });
      });
    });

    describe('when the Sidebar is closed', () => {
      beforeEach(() => {
        wrapper = mount(
          <Sidebar onCancel={ spy } />
        );
        instance = wrapper.instance();
        instance.listening = true;
      });

      it('removes event listeners for resize and closing', () => {
        spyOn(mockWindow, 'removeEventListener');
        wrapper.setProps({ title: 'Remove event handlers' });
        expect(mockWindow.removeEventListener.calls.count()).toEqual(1);
        expect(mockWindow.removeEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
      });
    });
  });

  describe('sidebarClasses', () => {
    it('returns a base sidebar', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('returns a base lefthand sidebar', () => {
      expect(leftWrapper).toMatchSnapshot();
    });
  });

  describe('render', () => {
    describe('when sidebar is closed', () => {
      beforeEach(() => {
        wrapper = shallow(
          <Sidebar
            onCancel={ spy }
            open={ false }
          >
          </Sidebar>
        );
      });

      it('sets all the correct classes', () => {
        expect(leftWrapper).toMatchSnapshot();
      });
    });
  });

  describe('Behaviour', () => {
    describe('clicking the close icon sidebar', () => {
      it('closes the sidebar', () => {
        let icon = portalContent.find('.carbon-sidebar__close-icon');
        icon.simulate('click');
        expect(spy).toHaveBeenCalled();
      });
    });
  });

  describe('tags', () => {
    beforeEach(() => {
      wrapper = shallow(
        <Sidebar
          open
          title='Test'
          onCancel={ () => {} }
          onConfirm={ () => {} }
          data-role='baz'
          data-element='bar'
        >
        </Sidebar>
      );
    });

    it('include correct component, element and role data tags', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
