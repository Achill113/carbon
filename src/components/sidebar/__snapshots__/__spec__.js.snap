// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar render when sidebar is closed sets all the correct classes 1`] = `
<Portal
  open={true}
>
  <div
    className="carbon-sidebar"
    data-component="sidebar"
  >
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="sidebar--left"
    >
      <div
        className="carbon-sidebar__sidebar carbon-sidebar__sidebar--left carbon-sidebar__sidebar--small"
        data-component="sidebar"
      />
    </CSSTransitionGroup>
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal-background"
    />
  </div>
</Portal>
`;

exports[`Sidebar sidebarClasses returns a base lefthand sidebar 1`] = `
<Portal
  open={true}
>
  <div
    className="carbon-sidebar"
    data-component="sidebar"
  >
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="sidebar--left"
    >
      <div
        className="carbon-sidebar__sidebar carbon-sidebar__sidebar--left carbon-sidebar__sidebar--small"
        data-component="sidebar"
      />
    </CSSTransitionGroup>
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal-background"
    />
  </div>
</Portal>
`;

exports[`Sidebar sidebarClasses returns a base sidebar 1`] = `
<Portal
  open={true}
>
  <div
    className="carbon-sidebar custom-class"
    data-component="sidebar"
  >
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="sidebar--right"
    >
      <div
        className="carbon-sidebar__sidebar carbon-sidebar__sidebar--right carbon-sidebar__sidebar--medium"
        data-component="sidebar"
      >
        <span
          className="carbon-sidebar__close"
        >
          <Icon
            bgSize="small"
            className="carbon-sidebar__close-icon"
            data-element="close"
            onClick={[Function]}
            type="close"
          />
        </span>
        <Textbox />
        <Textbox />
        <Textbox />
      </div>
    </CSSTransitionGroup>
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal-background"
    >
      <div
        className="carbon-modal__background"
      />
    </CSSTransitionGroup>
  </div>
</Portal>
`;

exports[`Sidebar tags include correct component, element and role data tags 1`] = `
<Portal
  open={true}
>
  <div
    className="carbon-sidebar"
    data-component="sidebar"
    data-element="bar"
    data-role="baz"
  >
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="sidebar--right"
    >
      <div
        className="carbon-sidebar__sidebar carbon-sidebar__sidebar--right carbon-sidebar__sidebar--medium"
        data-component="sidebar"
        data-element="bar"
        data-role="baz"
      >
        <span
          className="carbon-sidebar__close"
        >
          <Icon
            bgSize="small"
            className="carbon-sidebar__close-icon"
            data-element="close"
            onClick={[Function]}
            type="close"
          />
        </span>
      </div>
    </CSSTransitionGroup>
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal-background"
    >
      <div
        className="carbon-modal__background"
      />
    </CSSTransitionGroup>
  </div>
</Portal>
`;
